Dynamic Resource Allocation, devices, and drivers on Kubernetes blog post (#4631)
Azure/AKS: blog/_posts/2024-11-13-dra-devices-and-drivers-on-kubernetes.md

## Using GPUs in Kubernetes

Today, you need a few vendor-specific software components to even get started with a GPU node pool on Kubernetes - in particular, these are the Kubernetes device plugin and the GPU drivers. While the installation of GPU drivers makes sense, you might ask - why do we need a specific device plugin to be installed as well? 

Since Kubernetes does not have native support for special devices like GPUs, the device plugin surfaces and makes these resources available to your application. The device plugin works by exposing the number of GPUs on a node, by taking in a list of allocatable resources through the device plugin API and passing this to the kubelet. The kubelet then tracks this set and inputs the count of arbitrary resource type on the node to API Server, for kube-scheduler to use in pod scheduling decisions. 

![image](AKS/assets/images/dra-devices-drivers-on-kubernetes/device-plugin-diagram.png)

However, there are some limitations with this device plugin approach, namely:

1.	Granular selection of node resource type can only be made using several node attributes and labels

> 2.	Without a loopback mechanism, pods can be scheduled on “unprepared” nodes that aren’t ready to accept new resources

Author: ```suggestion
2.	Without a feedback loop between the device plugin and kube-scheduler (in which suitable nodes identified for a schedulable pod are also "checked" by the plugin based on factors like node readiness conditions), pods can be scheduled on “unprepared” nodes that aren’t ready to accept new resources
```

The Kubernetes scheduler will try multiple times to identify the right node and let the DRA resource driver(s) know that the node is ready for resource allocation.
Then the _resourceClaim_ stages record that the resource has been allocated on a particular node, and the scheduler is signaled to run the pod. (This prevents pods from being scheduled onto “unprepared” nodes that cannot accept new resources.)
Leveraging the Named Resources model, the DRA resource driver can specify:

1.	How to get a hold of a specific GPU type, and 
2.	How to configure the GPU that has been allocated to a pod and assigned to a node.
   
In the place of arbitrary resource count, an entire object now represents the choice of resource. This object is passed to the scheduler at node start time and may stream updates if resources become unhealthy.

Pulling this all together, the key components of DRA look like:

![image](AKS/assets/images/dra-devices-drivers-on-kubernetes/dra-driver-diagram.png)

## Seeing DRA in action

Now, let’s see how the [open-source k8s DRA driver](https://github.com/NVIDIA/k8s-dra-driver/tree/main) works with NVIDIA GPUs on a Kubernetes cluster, specifically how it can provide:

> * Exclusive access to a single GPU when multiple pods ask for it,

Author: ```suggestion
1.  Exclusive access to a single GPU when multiple pods ask for it,
```

> * Shared access within a pod with multiple containers, or

Author: ```suggestion
2.  Shared access within a pod with multiple containers, or
```

> * Shared access across pods requesting that single GPU.

Author: ```suggestion
3.  Shared access across pods requesting that single GPU.
```

> [!NOTE]

> > The following is an **experimental** demo on a local Kind cluster; the open-source k8s DRA resource driver is under active development and **not yet supported for production use** on Azure Kubernetes Service.

Author: ```suggestion
> The following is an **experimental** demo of the **first** scenario on a local Kind cluster; the open-source k8s DRA resource driver is under active development and **not yet supported for production use** on Azure Kubernetes Service.
```

```bash
kubectl get pod -A

---
NAMESPACE      NAME      READY      STATUS      RESTARTS    AGE
gpu-test1      pod1      1/1        Running     0           4s
gpu-test1      pod2      0/1        Pending     0           4s
...
...
---

kubectl describe pod -n gpu-test1 pod2 | less

---
Name:   pod2
...
Status: Pending
...
Conditions:
  Type            Status
  PodScheduled    False
Volumes: ...
...
Events:
  Type      Reason              Age                From                Message
  ----      ------              ---                ----                -------
  Warning   FailedScheduling    23s (x2 over 25s)  default-scheduler   0/2 nodes are available: 1 cannot allocate all claims...
---
```

Let's say we delete `pod1`. Now, `pod2` will get exclusive access to the single GPU:

```bash
kubectl delete pod -n gpu-test1 pod1
kubectl get pod -A

---
NAMESPACE      NAME      READY      STATUS      RESTARTS    AGE
gpu-test1      pod2      1/1        Running     0           56s
...
...
```

> Through this example and [many others](https://github.com/NVIDIA/k8s-dra-driver/tree/main), we see how the scheduler is told whether a GPU is fully committed to one workload or split across many workloads. 

Author: ```suggestion
Through this example and [many others](https://github.com/NVIDIA/k8s-dra-driver/tree/main), we see how the kube-scheduler is told whether a GPU is fully committed to one workload or split across many workloads. 
```

Leveraging the Named Resources model, the DRA resource driver can specify:

1.	How to get a hold of a specific GPU type, and 
2.	How to configure the GPU that has been allocated to a pod and assigned to a node.
   
In the place of arbitrary resource count, an entire object now represents the choice of resource. This object is passed to the scheduler at node start time and may stream updates if resources become unhealthy.

Pulling this all together, the key components of DRA look like:

![image](AKS/assets/images/dra-devices-drivers-on-kubernetes/dra-driver-diagram.png)

## Seeing DRA in action

Now, let’s see how the [open-source k8s DRA driver](https://github.com/NVIDIA/k8s-dra-driver/tree/main) works with NVIDIA GPUs on a Kubernetes cluster, specifically how it can provide:
* Exclusive access to a single GPU when multiple pods ask for it,
* Shared access within a pod with multiple containers, or
* Shared access across pods requesting that single GPU.

> [!NOTE]
> The following is an **experimental** demo on a local Kind cluster; the open-source k8s DRA resource driver is under active development and **not yet supported for production use** on Azure Kubernetes Service.

> <!-- Validating this demo on an AKS cluster currently, will replace in for 'local Kind cluster' in the above Note when completed. -->

Author: ```suggestion
```

The example above demonstrates GPU sharing within a pod, where two containers get access to one ResourceClaim object created for the pod.

The Kubernetes scheduler will try multiple times to identify the right node and let the DRA resource driver(s) know that the node is ready for resource allocation.
Then the _resourceClaim_ stages record that the resource has been allocated on a particular node, and the scheduler is signaled to run the pod. (This prevents pods from being scheduled onto “unprepared” nodes that cannot accept new resources.)
Leveraging the Named Resources model, the DRA resource driver can specify:

1.	How to get a hold of a specific GPU type, and 
2.	How to configure the GPU that has been allocated to a pod and assigned to a node.
   
In the place of arbitrary resource count, an entire object now represents the choice of resource. This object is passed to the scheduler at node start time and may stream updates if resources become unhealthy.

> Thus, one of the jobs of the DRA resource driver is to “translate” the resource vendor’s parameters into Resource Claim built-in parameters that have a defined type supported by Kubernetes. (This does not introduce any changes to CAS or its current performance.)

Author: ```suggestion
```

> 

Author: ```suggestion
```
